### PR TITLE
Feature: allow to define fragment outputs variables name in context

### PR DESCRIPTION
**Pull request by lciolecki**

### CODE REVIEW

File: ui/client/components/graph/node-modal/ParameterOutputList.tsx [hunk 1]
```diff
@@ -0,0 +1,119 @@
+import React, {useEffect, useState} from "react"
+import Field, {FieldType} from "./editors/field/Field"
+import {allValid, errorValidator, Validator} from "./editors/Validators"
+import {NodeType, NodeValidationError, ProcessDefinitionData} from "../../../types"
+import ProcessUtils from "../../../common/ProcessUtils"
+import {useDiffMark} from "./PathsToMark"
+import {useTranslation} from "react-i18next"
+
+type OutputFieldProps<N extends string, V> = {
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply ui/client/components/graph/node-modal/ParameterOutputList.tsx:9]: You're right we don't need N and V -> I've removed it. But class `OutputField`.

File: ui/client/components/graph/node-modal/ParameterOutputList.tsx [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    renderFieldLabel: (paramName: string) => JSX.Element,
+    setProperty: <K extends keyof NodeType>(property: K, newValue: NodeType[K], defaultValue?: NodeType[K]) => void,
+    fieldErrors?: NodeValidationError[],
+    showValidation?: boolean,
+    isEditMode?: boolean,
+}): JSX.Element {
+  const parameters = ProcessUtils.findNodeObjectTypeDefinition(editedNode, processDefinitionData.processDefinition)?.outputParameters
+  const [params, setParams] = useState(() => parameters.reduce((previousValue, currentValue) => ({
+    ...previousValue, [currentValue]: editedNode.ref?.outputParameters[currentValue],
+  }), {}))
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply ui/client/components/graph/node-modal/ParameterOutputList.tsx:79]: We're doing here reduce on `outputParameters` looking for already set param value in editedNode, base and base has to be `{}` to fill this object.

File: interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/SubprocessResolver.scala [hunk 3]
```diff
@@ -117,11 +117,16 @@ case class SubprocessResolver(subprocesses: String => Option[CanonicalProcess])
...
 
   //we replace outputs in subprocess with part of parent process
-  private def replaceCanonicalList(replacement: Map[String, CanonicalBranch], parentId: String): CanonicalBranch => ValidatedWithBranches[CanonicalBranch] = {
+  private def replaceCanonicalList(replacement: Map[String, CanonicalBranch], parentId: String, outputs: Option[Map[String, String]]): CanonicalBranch => ValidatedWithBranches[CanonicalBranch] = {
     iterateOverCanonicals({
-      case FlatNode(SubprocessOutputDefinition(id, name, fields, add)) => replacement.get(name) match {
-        case Some(nodes) => validBranches(FlatNode(SubprocessOutput(id, name, fields, add)) :: nodes)
-        case None => invalidBranches(UnknownSubprocessOutput(name, Set(id, parentId)))
+      case FlatNode(SubprocessOutputDefinition(id, name, fields, add)) => {
+        val outputName = outputs.flatMap(_.get(name)).getOrElse(name)
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/SubprocessResolver.scala]: Done

File: ui/server/src/main/scala/pl/touk/nussknacker/ui/definition/UIProcessObjectsFactory.scala [hunk 4]
```diff
@@ -113,12 +113,20 @@ object UIProcessObjectsFactory {
...
-      case SubprocessDetails(CanonicalProcess(MetaData(id, FragmentSpecificData(docsUrl), _, _), FlatNode(SubprocessInputDefinition(_, parameters, _)) :: _, _), category) =>
+      case fragment@SubprocessDetails(CanonicalProcess(MetaData(id, FragmentSpecificData(docsUrl), _, _), FlatNode(SubprocessInputDefinition(_, parameters, _)) :: _, _), category) =>
         val config = fixedComponentsConfig.getOrElse(id, SingleComponentConfig.zero).copy(docsUrl = docsUrl)
         val typedParameters = parameters.map(extractSubprocessParam(classLoader, config))
-        (id, new ObjectDefinition(typedParameters, Typed[java.util.Map[String, Any]], Some(List(category)), config))
+
+
+        //Figure outputs parameter
+        val outputParameters = fragment.canonical.collectAllNodes.collect {
+          case SubprocessOutputDefinition(_, name, fields, _) if fields.nonEmpty => name
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply ui/server/src/main/scala/pl/touk/nussknacker/ui/definition/UIProcessObjectsFactory.scala:125]: Done

File: interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/SubprocessResolver.scala [hunk 5]
```diff
@@ -117,11 +117,21 @@ case class SubprocessResolver(subprocesses: String => Option[CanonicalProcess])
...
 
   //we replace outputs in subprocess with part of parent process
-  private def replaceCanonicalList(replacement: Map[String, CanonicalBranch], parentId: String): CanonicalBranch => ValidatedWithBranches[CanonicalBranch] = {
+  private def replaceCanonicalList(replacement: Map[String, CanonicalBranch], parentId: String, outputs: Option[Map[String, String]]): CanonicalBranch => ValidatedWithBranches[CanonicalBranch] = {
     iterateOverCanonicals({
-      case FlatNode(SubprocessOutputDefinition(id, name, fields, add)) => replacement.get(name) match {
-        case Some(nodes) => validBranches(FlatNode(SubprocessOutput(id, name, fields, add)) :: nodes)
-        case None => invalidBranches(UnknownSubprocessOutput(name, Set(id, parentId)))
+      case FlatNode(SubprocessOutputDefinition(id, name, fields, add)) => {
+        val maybeOutputName = outputs match {
```
Comment (mproch) [inline interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/SubprocessResolver.scala:123]: let's add comment it's temporary

**Review by mproch — APPROVED** [review]
:+1: hope it will be good basis for further development ;)

**Comment by lciolecki** [comment]
> +1 hope it will be good basis for further development ;)

I think we will rewrite it.. ;)

**Pull request merged**